[R2] S3 API compatibility page (#4025)
cloudflare/cloudflare-docs: content/r2/platform/S3 Compatibility/s3-compatibility.md

---
title: S3 API Compatibility
pcx-content-type: reference
meta:
  title: S3 API Compatibility
---

# S3 API compatibility

Feature implementation is currently in progress. Refer back to this page for updates.

## How to read this page

> Each operations section will include a table of implemented and a table of unimplemented APIs. Refer the feature column of each table to review which specific features of the API have been implemented and which are not.

Reviewer: ```suggestion
Each operations section will include two tables: one of implemented APIs and one of unimplemented APIs. Refer the feature column of each table to review which specific features of the API have been implemented and which have not.
```

✅ Feature Implemented  
❌ Feature Not Implemented

## Bucket-level operations

Below is a list of implemented bucket-level operations. Refer to the Feature column to review which features have been implemented (✅) and have not been implemented (❌).

> 

Reviewer: Is this meant to be the implemented table? If so, label it. If not, should be combined with table below.

{{<table-wrap style="width:100%">}}

| API Name                                                                                | Feature                           |
| -------------------------------------------------------------------------------------   |---------------------------------- |

> | [ListBuckets](https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListBuckets.html)     |                                    |

Reviewer: Sorta confused why we have this command but nothing to say about it?

| [HeadBucket](https://docs.aws.amazon.com/AmazonS3/latest/API/API_HeadBucket.html)       | ❌ Bucket Owner: <br> ❌ x-amz-expected-bucket-owner  |

> | [CreateBucket](https://docs.aws.amazon.com/AmazonS3/latest/API/API_CreateBucket.html)   | ❌ ACL: <br> ❌ x-amz-acl <br> ❌ x-amz-grant-full-control <br> ❌ x-amz-grant-read <br> ❌ x-amz-grant-read-acp  <br> ❌ x-amz-grant-write <br> ❌ x-amz-grant-write-acp |

Reviewer: Not sure I'm in love with the single column formatting for this (sorta hard to tell which is a top-level feature and which isn't).

Don't really know if we have too too many other options for formatting though?

Reviewer: I think it's mostly just confusing having a check next to the category name (`Operation Metadata`, `System Metadata`) in addition to the sub-items (especially when they're the same level of indentation).

|                                                                                         | ❌ SSE-C: <br> ❌ x-amz-server-side-encryption <br> ❌ x-amz-server-side-encryption-customer-algorithm <br> ❌ x-amz-server-side-encryption-customer-key <br> ❌ x-amz-server-side-encryption-customer-key-MD5 <br> ❌ x-amz-server-side-encryption-aws-kms-key-id <br> ❌ x-amz-server-side-encryption-context <br> ❌ x-amz-server-side-encryption-bucket-key-enabled |
|                                                                                         | ❌ Request Payer: <br> ❌ x-amz-request-payer |
|                                                                                         | ❌ Tagging: <br> ❌ x-amz-tagging |      |
|                                                                                         | ❌ Object Locking: <br> ❌ x-amz-object-lock-mode <br> ❌ x-amz-object-lock-retain-until-date <br> ❌ x-amz-object-lock-legal-hold |
|                                                                                         | ❌ ACL: <br> ❌ x-amz-acl <br> ❌ x-amz-grant-full-control <br> ❌ x-amz-grant-read <br> ❌ x-amz-grant-read-acp <br> ❌ x-amz-grant-write-acp |

> |                                                                                         | ❌ Bucket Owner: <br> ❌ x-amz-expected-bucket-owner |

Reviewer: Storage class?
